### PR TITLE
Support Bigscape v2 data

### PR DESCRIPTION
**Pull request by adraismawur**
This PR enables the user to load data from a BiG-SCAPE V2 database file. The `data_sqlite.db` file is expected to be in the same place as the `mix_clustering_[cutoff].tsv` would otherwise be. 

If both files are found, the loader selects the tsv file.

### CODE REVIEW

File: src/nplinker/arranger.py [hunk 1]
```diff
@@ -455,5 +455,6 @@ def validate_bigscape(bigscape_dir: Path) -> None:
         raise FileNotFoundError(f"BiG-SCAPE data directory not found at {bigscape_dir}")
 
     clustering_file = bigscape_dir / f"mix_clustering_c{config.bigscape.cutoff}.tsv"
-    if not clustering_file.exists():
-        raise FileNotFoundError(f"BiG-SCAPE clustering file not found: {clustering_file}")
+    database_file = bigscape_dir / "data_sqlite.db"
```
Comment (CunliangGeng) [inline src/nplinker/arranger.py:482]: The name `data_sqlite.db` is pretty generic, and I suggest adding a prefix e.g. `bigscape_data.db`

Comment (CunliangGeng) [inline src/nplinker/arranger.py:482]: Please update the docstrings of methods `validate_bigscape` and `arrange_bigscape` to add the requirements on this database file

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import sqlite3
+from os import PathLike
+from nplinker.logconfig import LogConfig
+from ..abc import GCFLoaderBase
+from ..gcf import GCF
+
+
+logger = LogConfig.getLogger(__name__)
+
+class BigscapeV2GCFLoader:
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]: Move this class to the file `bigscape_loader.py`?

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    def __init__(self, db_file: str | PathLike, /) -> None:
+        """Build a loader for BiG-SCAPE v2 database file.
+
+        Args:
+            cluster_file: Path to the BiG-SCAPE v2database file,
+                the filename has a pattern of "data_sqlite.db".
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]:
```suggestion
            db_file: Path to the BiG-SCAPE v2 database file,
                the filename has a pattern of "bigscape_data.db".
```

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        if not keep_mibig_only:
+            gcf_list = [gcf for gcf in gcf_list if not gcf.has_mibig_only()]
+        if not keep_singleton:
+            gcf_list = [gcf for gcf in gcf_list if not gcf.is_singleton()]
+        return gcf_list
+
+
+    @staticmethod
+    def _parse_gcf(db_file: str) -> list[GCF]:
+        """Get GCF objects from database
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]:
```suggestion
        """Get GCF objects from database.
```

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    def __init__(self, db_file: str | PathLike, /) -> None:
+        """Build a loader for BiG-SCAPE v2 database file.
+
+        Args:
+            cluster_file: Path to the BiG-SCAPE v2database file,
+                the filename has a pattern of "data_sqlite.db".
+
+        Attributes:
+            cluster_file: path to the BiG-SCAPE database file.
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]:
```suggestion
            db_file: path to the BiG-SCAPE database file.
```

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py (hunk 3 above)
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]: Is it still necessary to mention the filename pattern?

File: src/nplinker/loader.py [hunk 6]
```diff
@@ -159,7 +159,24 @@ def _load_genomics(self):
         bigscape_cluster_file = (
             globals.BIGSCAPE_DEFAULT_PATH / f"mix_clustering_c{config.bigscape.cutoff}.tsv"
         )
-        raw_gcfs = BigscapeGCFLoader(bigscape_cluster_file).get_gcfs()
+        bigscape_db_file = (
+            globals.BIGSCAPE_DEFAULT_PATH / f"data_sqlite.db"
+        )
+
+        # switch depending on found file. prefer V1 if both are found
```
Comment (CunliangGeng) [inline src/nplinker/loader.py:165]: It's ok to use this logics at the moment. Later it should be adapted e.g. with a new config on bigscape version.

File: tests/unit/genomics/test_bigscape_loader_v2.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import pytest
+from nplinker.genomics import GCF
+from nplinker.genomics.abc import GCFLoaderBase
+from nplinker.genomics.bigscape import BigscapeV2GCFLoader
+from .. import DATA_DIR
+
+
+class TestBigscapelGCFLoader:
```
Comment (CunliangGeng) [inline tests/unit/genomics/test_bigscape_loader_v2.py]: Same, move this class to `test_bigscape_loader.py`?

**Review by CunliangGeng — CHANGES_REQUESTED** [review]
Thanks for adding the support of bigscape v2!

I left some minor comments for you.

One final comment on all files is that please remember to format all files using `ruff`. If you are also using vscode, the formatting could be automatic, see the [guide](https://github.com/NPLinker/nplinker/blob/dev/README.dev.md#linting-and-formatting).

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        """Get all GCF objects.
+
+        Args:
+            keep_mibig_only: True to keep GCFs that contain only MIBiG
+                BGCs.
+            keep_singleton: True to keep singleton GCFs. A singleton GCF
+                is a GCF that contains only one BGC.
+
+        Returns:
+            list[GCF]: a list of GCF objects.
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]:
```suggestion
            A list of GCF objects.
```

Comment (CunliangGeng) [reply src/nplinker/genomics/bigscape/bigscape_loader_v2.py]: The static typing is not needed in docstring and the [API doc](https://nplinker.github.io/nplinker/latest/api/bigscape/) will add it automatically.

File: src/nplinker/genomics/bigscape/bigscape_loader_v2.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    @staticmethod
+    def _parse_gcf(db_file: str) -> list[GCF]:
+        """Get GCF objects from database
+
+        Args:
+            db_file: Path to the sqlite3 database file.
+
+        Returns:
+            list: A list of GCF objects
```
Comment (CunliangGeng) [inline src/nplinker/genomics/bigscape/bigscape_loader_v2.py]:
```suggestion
            A list of GCF objects
```

File: src/nplinker/arranger.py (hunk 1 above)
Comment (adraismawur) [reply src/nplinker/arranger.py:482]: Indeed! This is what the output is named after BiG-SCAPE v2 runs, but we can rename it to something more descriptive.

Comment (CunliangGeng) [reply src/nplinker/arranger.py:482]: Aha, then better to keep the same name in NPLinker, we don't want to bother users to rename it. If bigscape changes this name, then nplinker changes as well.

**Comment by CunliangGeng** [comment]
Just [merged a PR](https://github.com/NPLinker/nplinker/pull/235) to ignore docstrings for test files. So it's ok to not have docstrings for test functions.

I'm off from now, @gcroci2 could take over this PR.

**Comment by gcroci2** [comment]
> Just [merged a PR](https://github.com/NPLinker/nplinker/pull/235) to ignore docstrings for test files. So it's ok to not have docstrings for test functions.
> 
> I'm off from now, @gcroci2 could take over this PR.

Sure! @adraismawur just ask for my review again when you'll have integrated Cunliang's one :)

**Comment by adraismawur** [comment]
Alright! Changes were committed. Could you review this again?

File: src/nplinker/loader.py [hunk 10]
```diff
@@ -159,7 +159,21 @@ def _load_genomics(self):
...
+        if bigscape_cluster_file.exists():
+            loader = BigscapeGCFLoader(bigscape_cluster_file)
+            logger.debug(f"Loading BigSCAPE cluster file {bigscape_cluster_file}")
+        elif bigscape_db_file.exists():
+            loader = BigscapeV2GCFLoader(bigscape_db_file)
+            logger.debug(f"Loading BigSCAPE database file {bigscape_db_file}")
+        else:
+            raise FileNotFoundError(
+                f"BigSCAPE cluster file {bigscape_cluster_file} or database file {bigscape_db_file} not found."
```
Comment (gcroci2) [inline src/nplinker/loader.py]:
```suggestion
                f"Neither BigSCAPE cluster file {bigscape_cluster_file} nor database file {bigscape_db_file} were found."
```

**Review by gcroci2 — CHANGES_REQUESTED** [review]
All good :) Thanks for the contribution!

The action for checking the format needs to pass before merging. Please run ruff as explained [here](https://github.com/NPLinker/nplinker/blob/dev/README.dev.md#linting-and-formatting) and then commit the format/linting fixes.

**Comment by adraismawur** [comment]
Ok I have run the formatter, but there are files outside of those I edited in this PR that fail using ruff.

Should I rebase dev onto this branch and add formatting changes of other files to this one, or should I just merge as-is and the formatting can be done elsewhere?

**Comment by gcroci2** [comment]
> Ok I have run the formatter, but there are files outside of those I edited in this PR that fail using ruff.
> 
> Should I rebase dev onto this branch and add formatting changes of other files to this one, or should I just merge as-is and the formatting can be done elsewhere?

I still see:
`src/nplinker/arranger.py:231:101: E501 Line too long (114 > 100)`
Please make sure that the ruff checks pass on the files you edited (only) and then feel free to merge to dev. No worries about other files you didn't touch with this PR. 

About the linting check failing, is this expected? @CunliangGeng If yes, I think we should exclude the checking on such files on the CI, otherwise it can be confusing for contributors.

**Review by gcroci2 — APPROVED** [review]

**Comment by CunliangGeng** [comment]
@gcroci2 @adraismawur 

> About the linting check failing, is this expected? @CunliangGeng If yes, I think we should exclude the checking on such files on the CI, otherwise it can be confusing for contributors.

Good point! I updated the github actions to only check changed files in PR #238.

**Pull request merged**